src/components/NavBar/ChainSwitcher.tsx

```diff
@@ -0,0 +1,178 @@
+    return null
+  }
+  const active = chainId === targetChain
+  const { label, logoUrl } = getChainInfo(targetChain)
+
+  return (
+    <Row
+      as="button"
+      background={active ? 'lightGrayContainer' : 'none'}
+      className={`${styles.ChainSwitcherRow} ${subhead}`}
+      onClick={() => onSelectChain(targetChain)}
+    >
+      <ChainDetails>
+        <img src={logoUrl} alt={label} className={styles.Icon} />
+        {label}
+      </ChainDetails>
+      {active && <CheckMarkIcon width={20} height={20} />}
+    </Row>
+  )
+}
+
+const ChainDetails = ({ children }: { children: ReactNode }) => <Row>{children}</Row>
+
+const getParsedChainId = (parsedQs?: ParsedQs) => {
+  const chain = parsedQs?.chain
+  if (!chain || typeof chain !== 'string') return { urlChain: undefined, urlChainId: undefined }
+
+  return { urlChain: chain.toLowerCase(), urlChainId: getChainIdFromName(chain) }
+}
+
+const getChainIdFromName = (name: string) => {
+  const entry = Object.entries(CHAIN_IDS_TO_NAMES).find(([_, n]) => n === name)
+  const chainId = entry?.[0]
+  return chainId ? parseInt(chainId) : undefined
+}
+
+const getChainNameFromId = (id: string | number) => {
+  // casting here may not be right but fine to return undefined if it's not a supported chain ID
+  return CHAIN_IDS_TO_NAMES[id as SupportedChainId] || ''
+}
+
+const NETWORK_SELECTOR_CHAINS = [
+  SupportedChainId.MAINNET,
+  SupportedChainId.POLYGON,
+  SupportedChainId.OPTIMISM,
+  SupportedChainId.ARBITRUM_ONE,
+]
+
+interface ChainSwitcherProps {
+  isMobile?: boolean
+}
+
+export const ChainSwitcher = ({ isMobile }: ChainSwitcherProps) => {
+  const dispatch = useAppDispatch()
+  const { chainId, provider, connector } = useWeb3React()
+  const previousChainId = usePrevious(chainId)
+  const [isOpen, toggleOpen] = useReducer((s) => !s, false)
+  const parsedQs = useParsedQueryString()
+  const { urlChain, urlChainId } = getParsedChainId(parsedQs)
+  const previousUrlChainId = usePrevious(urlChainId)
+
+  const ref = useRef<HTMLDivElement>(null)
+  useOnClickOutside(ref, isOpen ? toggleOpen : undefined)
+
+  const navigate = useNavigate()
+  const location = useLocation()
+  const info = chainId ? getChainInfo(chainId) : undefined
+
+  const onSelectChain = useCallback(
+    async (targetChain: number, skipToggle?: boolean) => {
+      if (!connector) return
+
+      const connectionType = getConnection(connector).type
+
+      try {
+        dispatch(updateConnectionError({ connectionType, error: undefined }))
+        await switchChain(connector, targetChain)
+      } catch (error) {
+        console.error('Failed to switch networks', error)
+        dispatch(updateConnectionError({ connectionType, error: error.message }))
+        dispatch(addPopup({ content: { failedSwitchNetwork: targetChain }, key: `failed-network-switch` }))
+      }
+
+      if (!skipToggle) {
+        toggleOpen()
+      }
+    },
+    [connector, toggleOpen, dispatch]
+  )
+
+  useEffect(() => {
```

Comment:
i think we should move these into `App.tsx` instead, so then you won't need to re-do them here.